EIPS/blockhash_refactoring.md

```diff
@@ -0,0 +1,84 @@
+### Preamble
+
+    EIP: <to be assigned>
+    Title: Blockhash refactoring
+    Author: Vitalik Buterin
+    Type: Standard Track
+    Category: Core
+    Status: Draft
+    Created: 2017-02-10
+    
+### Summary
+
+Stores blockhashes in the state, reducing the protocol complexity and the need for client implementation complexity in order to process the BLOCKHASH opcode. Also extends the range of how far back blockhash checking can go, with the side effect of creating direct links between blocks with very distant block numbers, facilitating much more efficient initial light client syncing.
+
+### Parameters
+
+* `METROPOLIS_FORK_BLKNUM`: TBD
+* `SUPER_USER`: 2**160 - 2
+* `BLOCKHASH_CONTRACT_ADDR`: 0xf0 (ie. 240)
+* `BLOCKHASH_CONTRACT_CODE`: see below
+
+### Specification
+
+If `block.number == METROPOLIS_FORK_BLKNUM`, then when processing the block, before processing any transactions set the code of BLOCKHASH_CONTRACT_ADDR to BLOCKHASH_CONTRACT_CODE.
+
+If `block.number >= METROPOLIS_FORK_BLKNUM`, then when processing a block, before processing any transactions execute a call with the parameters:
+
+* `SENDER`: SUPER_USER
+* `GAS`: 1000000
+* `TO`: BLOCKHASH_CONTRACT_ADDR
+* `VALUE`: 0
+* `DATA`: <32 bytes corresponding to the block's prevhash>
+
+If `block.number >= METROPOLIS_FORK_BLKNUM + 256`, then the BLOCKHASH opcode instead returns the result of executing a call with the parameters:
+
+* `SENDER`: <account from which the opcode was called>
+* `GAS`: 1000000
+* `TO`: BLOCKHASH_CONTRACT_ADDR
+* `VALUE`: 0
+* `DATA`: 32 byte zero-byte-leftpadded integer representing the stack argument with which the opcode was called
+
+Also, the gas cost is increased from 20 to 800 to reflect the higher costs of processing the algorithm in the contract code.
```

Comment:
I think it would make sense setting the cost to the upper bound of what the read actually costs.

If the current proposed rules make that upper bound very large and therefore pose an unrealistically high cost for the opcode, then I would suggest:
-  have a contract which stores less and has a lower higher bound *or*
- limiting the parameter to 256 (as currently) for the opcode